### DISCUSSION TITLE
Noah-MP MMF Crashes in HRLDAS v4.4

### DISCUSSION THREAD

**Discussion by TimLahmers** [NCAR/hrldas#193] [Q&A]
I am attempting to run Noah-MP HRDAS with the Miguez-Macho and Fan (MMF) scheme (Runoff Option 5) and the code is crashing without a useful stack trace. To run the model, I derived geo_em.nc and wrfinput.nc files using WPS and WRF v4.4, following the instructions provided by the HRLDAS team here: https://github.com/NCAR/hrldas/discussions/97

I am using HRLDAS v4.4 to stay consistent with earlier Noah-MP versions.

Are there other missing settings in HRLDAS v4.4 or is there a way to get a useful stack trace? I am running the code on NASA NCCS Discover (in the new SLES15 environment).

Model output is below.
    Timing:   0.29 Cumulative:        0.29  SFLX:   0.00
forrtl: error (72): floating overflow
Image              PC                Routine            Line        Source             
libpthread-2.31.s  000014691263F910  Unknown               Unknown  Unknown
hrldas.exe         00000000004D5AEE  Unknown               Unknown  Unknown
hrldas.exe         00000000004CE815  Unknown               Unknown  Unknown
hrldas.exe         00000000004B4073  Unknown               Unknown  Unknown
hrldas.exe         0000000000475729  Unknown               Unknown  Unknown
hrldas.exe         0000000000648867  Unknown               Unknown  Unknown
hrldas.exe         000000000040D3D2  Unknown               Unknown  Unknown
hrldas.exe         000000000040D34D  Unknown               Unknown  Unknown
libc-2.31.so       000014691231424D  __libc_start_main     Unknown  Unknown
hrldas.exe         000000000040D27A  Unknown               Unknown  Unknown
Aborted

**Comment by cenlinhe** [comment]
@CharlesZheZhang @prasanthvkrishna Could you please take a look? One note is that in this version, HRLDAS-MMF cannot be run in a parallel mode if I remember correctly.

**Comment by TimLahmers** [comment]
Thanks @cenlinhe I forgot about that. I was testing the model both in parallel and on a single node, but I originally built the model using option 4 (4. Linux ifort compiler MPI ). Would calling the model with MMF after building it in parallel result in a crash? If so, I can rebuild with option 3 (3. Linux ifort compiler serial) and try again.

**Reply by cenlinhe** [reply]
I think you may want to try to rebuild it with option 3, because once you built it with MPI option, it will use/activate the MPP HRLDAS packages. (although I am not sure if this is the root cause for your problem).

**Reply by TimLahmers** [reply]
Thanks for the suggestions, I rebuilt the model with option 3 and added some fortran flags to enable a stack trace:
    Timing:   0.32 Cumulative:        0.32  SFLX:   0.00
forrtl: error (72): floating overflow
Image              PC                Routine            Line        Source             
libpthread-2.31.s  0000148B684A1910  Unknown               Unknown  Unknown
hrldas.exe         00000000004D018E  module_sf_noahmpl        7465  module_sf_noahmplsm.f90
hrldas.exe         00000000004C8EB5  module_sf_noahmpl        6200  module_sf_noahmplsm.f90
hrldas.exe         00000000004AE713  module_sf_noahmpl         993  module_sf_noahmplsm.f90
hrldas.exe         000000000046FDC9  module_sf_noahmpd        1151  module_sf_noahmpdrv.f90
hrldas.exe         0000000000642E15  module_noahmp_hrl        2463  module_NoahMP_hrldas_driver.f90
hrldas.exe         000000000040C7D2  MAIN__                     24  main_hrldas_driver.f90
hrldas.exe         000000000040C74D  Unknown               Unknown  Unknown
libc-2.31.so       0000148B682C924D  __libc_start_main     Unknown  Unknown
hrldas.exe         000000000040C67A  Unknown               Unknown  Unknown

The line causing the crash is as follows:
RUNSRF = RUNSRF * 1000.0 + RSAT * 1000.0/DT  ! m/s -> mm/s

From adding a few print statements, RSAT is unphysical:
RUNSRF:   0.0000000E+00
RSAT:     9.9692093E+36
DT:        3600.000    

RSAT is derived from WPLUS (an output from SSTEP). Is there a Noah-MP variable that can cause this? It looks like a no-data value from a parameter file went into a calculation.

**Comment by CharlesZheZhang** [comment]
Hi Tim, 
RSAT: 9.9692093E+36 means a nan value that could come from SSTEP subroutine:
https://github.com/NCAR/noahmp/blob/9f4feb02ddc868303b8d3590d2a3aba0aebbde94/src/module_sf_noahmplsm.F#L7918
with OPT_RUN=5.
Please make sure variables in this calculation have reasonable values, especially SMCWTD, which is needed in calculating WPLUS.

**Reply by RAbolafiaRosenzweig** [reply]
Agreed here, it may also be worth making sure the soil classification inputs are all within the accepted range of values.

**Reply by TimLahmers** [reply]
Thanks for your feedback on this. It looks like the MMF variables going into HRLDAS are missing in the geo_em.nc and are set to zero in the wrfinput.nc file. This suggests an issue with when I ran WPS. I will need to check my configuration.

**Reply by cenlinhe** [reply]
When you run WPS, please use the GEOGRID.TBL.ARW_noahmp look-up table instead of the default GEOGRID.TBL.ARW.

**Reply by TimLahmers** [reply]
Thanks Cenlin, that looks like the issue. I was using the wrong GEOGRID.TBL

**Reply by TimLahmers** [reply]
I have one more question on this. Updating the GEOGRID.TBL fixes the geo_em.nc file; however, the wrfinput_d01 file (based on the geogrid) still sets the MMF variables to zeros. I am generating the file using real.exe. Is there a setting to make sure those variables get copied into the final wrfinput_d01 netcdf file?

**Reply by CharlesZheZhang** [reply]
When running real.exe, you also need to set opt_run=5 in namelist.input, this will run the groundwater_init subroutine to initialize groundwater and soil moisture fields related to MMF scheme.

**Reply by TimLahmers** [reply]
Thanks, I just checked and opt_run is set to 5. Are there any other possible causes for blank variables in the wrfinput_d01 files?
